File: projects/RandoMainDLL/PickupCollection.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Globalization;
using System.Linq;
using System.Text.RegularExpressions;
using RandoMainDLL.Memory;

namespace RandoMainDLL {
  using Conditions = SortedDictionary<UberStateCondition, Pickup>;
  using States = Dictionary<UberId, SortedDictionary<UberStateCondition, Pickup>>;



  public class PickupDictionary {
    // Optimize the equals pickup as we only match one of those on collection.
    Dictionary<UberStateCondition, Pickup> UberStateEqualsPickups = new Dictionary<UberStateCondition, Pickup>();
    States UberStatePickups = new States();

    private Conditions GetConditions(UberId id) {
      if (UberStatePickups.TryGetValue(id, out var pickups))
        return pickups;

      var dict = new Conditions();
      UberStatePickups.Add(id, dict);
      return dict;
    }

    private Tuple<UberStateCondition, Pickup> GetEqCondition(UberId id, double value) {
      var cond = new UberStateCondition(id, (int)value, UberStateCondition.Handler.Equals);
      return Tuple.Create(cond, UberStateEqualsPickups.GetOrElse(cond, null));
    }

    private Tuple<UberStateCondition, Pickup> GetEqCondition(UberState state) {
      var cond = new UberStateCondition(state.GroupID, state.ID, state.ValueAsInt(), UberStateCondition.Handler.Equals);
      return Tuple.Create(cond, UberStateEqualsPickups.GetOrElse(cond, null));
    }

    public Tuple<UberStateCondition, Pickup> PickupWithCondition(UberId id, double value) {
      var conditions = GetConditions(id);
      foreach (var condition in conditions)
        if (condition.Key.Met(value))
          return Tuple.Create(condition.Key, condition.Value);

      var eq = GetEqCondition(id, value);
      if (eq.Item2 != null)
        return eq;

      return Tuple.Create<UberStateCondition, Pickup>(null, Multi.Empty);
    }

    public Pickup Pickup(UberStateCondition cond) {
      if (cond.TargetHandler == UberStateCondition.Handler.Equals)
        return UberStateEqualsPickups.GetOrElse(cond, Multi.Empty);
      else
        return UberStatePickups.GetOrElse(cond.Id, null).GetOrElse(cond, Multi.Empty);
    }

    public Pickup Pickup(UberId id, double value) {
      return PickupWithCondition(id, value).Item2;
    }

    public void Clear() {
      UberStateEqualsPickups.Clear();
      UberStatePickups.Clear();
    }

    public bool HasPickup(UberId id, double value) {
      return PickupWithCondition(id, value).Item1 != null;
    }

    public bool Collect(UberState state, UberValue value) {
      var collected = false;
      var conditions = GetConditions(state.GetUberId());
      foreach (var condition in conditions)
        if (condition.Key.Met(state, value))
          collected |= condition.Value.Collect(condition.Key);

      var eq = GetEqCondition(state);
      if (eq.Item2 != null)
        collected |= eq.Item2.Collect(eq.Item1);

      return collected;
    }

    public void Add(UberStateCondition condition, Pickup pickup) {
      if (condition.TargetHandler == UberStateCondition.Handler.Equals) {
        var p = UberStateEqualsPickups.GetOrElse(condition, Multi.Empty);
        UberStateEqualsPickups.Remove(condition);
        UberStateEqualsPickups.Add(condition, p.Concat(pickup));
      }
      else {
        var conditions = GetConditions(condition.Id);
        var p = conditions.GetOrElse(condition, Multi.Empty);
        conditions.Remove(condition);
        conditions.Add(condition, p.Concat(pickup));
      }
    }

    public int PickupCount(ZoneType type) {
      return UberStatePickups.Aggregate(0, (s, d) => s + d.Value.Count(p => p.Key.Loc().Zone == type))
        + UberStateEqualsPickups.Count(p => p.Key.Loc().Zone == type);
    }
  }
}

File: projects/Randomizer/pickups/pickup_collection.cpp
#include <randomizer/pickups/pickup_collection.h>

namespace randomizer {

    ZoneType get_zone(uber_states::UberStateCondition location) {
        return ZoneType::Void;
    }

    PickupCollection::conditioned_pickup PickupCollection::get_eq_condition(uber_states::UberState state, double value) {
        uber_states::UberStateCondition cond{ state, uber_states::UberStateCondition::Operator::Equals, value };
        auto it = uber_state_equals_pickups.find(cond);
        if (it == uber_state_equals_pickups.end())
            return conditioned_pickup();

        return std::make_pair(cond, it->second);
    }

    PickupCollection::conditioned_pickup PickupCollection::get_eq_condition(uber_states::UberState state) {
        return get_eq_condition(state, state.get());
    }

    PickupCollection::conditioned_pickup PickupCollection::pickup_with_condition(uber_states::UberState state, double value) {
        auto conditions = uber_state_pickups[state];
        for (auto condition : conditions)
            if (condition.first.resolve(value))
                return condition;

        return get_eq_condition(state, value);
    }

    PickupList PickupCollection::pickup(uber_states::UberStateCondition cond) {
        if (cond.op != uber_states::UberStateCondition::Operator::Equals) {
            auto it_conditions = uber_state_pickups.find(cond.state);
            if (it_conditions == uber_state_pickups.end())
                return PickupList();

            auto it_pickups = it_conditions->second.find(cond);
            return it_pickups != it_conditions->second.end() ? it_pickups->second : PickupList();
        }

        auto it = uber_state_equals_pickups.find(cond);
        return it != uber_state_equals_pickups.end() ? it->second : PickupList();
    }

    PickupList PickupCollection::pickup(uber_states::UberState state, double value) {
        auto pickup = pickup_with_condition(state, value);
        return pickup.has_value() ? pickup.value().second : PickupList();
    }

    void PickupCollection::clear() {
        uber_state_equals_pickups.clear();
        uber_state_pickups.clear();
    }

    bool PickupCollection::has_pickup(uber_states::UberState state, double value) {
        auto pickup = pickup_with_condition(state, value);
        return pickup.has_value() && !pickup.value().second.empty();
    }

    bool PickupCollection::collect(uber_states::UberState state, double value) {
        auto collected = false;
        auto conditions = uber_state_pickups[state];
        for (auto condition : conditions)
            if (condition.first.resolve(state, value))
                collected |= condition.second.collect(condition.first);

        auto eq = get_eq_condition(state);
        if (eq.has_value())
            collected |= eq.value().second.collect(eq.value().first);

        return collected;
    }

    void PickupCollection::add(uber_states::UberStateCondition condition, std::shared_ptr<Pickup>&& pickup) {
        if (condition.op == uber_states::UberStateCondition::Operator::Equals)
            uber_state_equals_pickups[condition].add(std::move(pickup));
        else
            uber_state_pickups[condition.state][condition].add(std::move(pickup));
    }

    int PickupCollection::pickup_count(ZoneType zone) {
        auto count = std::accumulate(uber_state_pickups.begin(), uber_state_pickups.end(), 0, [zone](int sum, states::value_type const& value) {
            return sum + static_cast<int>(std::count_if(value.second.begin(), value.second.end(), [zone](conditions::value_type const& pickup) {
                return !pickup.second.empty() && get_zone(pickup.first) == zone;
            }));
        });
        
        count += static_cast<int>(std::count_if(uber_state_equals_pickups.begin(), uber_state_equals_pickups.end(), [zone](conditions::value_type const& pickup) {
            return !pickup.second.empty() && get_zone(pickup.first) == zone;
        }));

        return 0;
    }
} // namespace randomizer
